### PR TITLE
Compressed layout moves suptitle

### LINKED ISSUE #27866: [Bug]: Adding suptitle in compressed layout causes weird spacing
### Bug summary

If I create a "compressed" layout with two rows with both subplots having fixed aspect ratios, adding a suptitle will insert the space between the suptitle and the plots. I would expect the excess space to be put in below the subplots.

### Code for reproduction

```Python
fig, (ax0, ax1) = plt.subplots(
    nrows=2, figsize=(4, 10), layout="compressed",
    gridspec_kw={"height_ratios": (1 / 4, 3 / 4), "hspace": 0}
)

ax0.axis("equal")
ax0.set_box_aspect(1/3)

ax1.axis("equal")
ax1.set_box_aspect(1)

# Adding suptitle causes weird spacing
fig.suptitle("Title")
```

### Actual outcome

With suptitle:

image

Without suptitle:

image

Notably, this also happens without defining the gridspec kwargs.

### Expected outcome

I would expect the extra spacing to be added at the bottom, not be inserted between the suptitle and the plots.

### Additional information

_No response_

### Operating system

OSX

### Matplotlib Version

3.8.0

### Matplotlib Backend

module://matplotlib_inline.backend_inline

I am running this inside jupyter lab, but this persists even if I save the figure to a file.

### Python version

_No response_

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by rcomer**
## PR summary

Closes #27866.  I found myself wanting this today.  The code from the issue now produces

test

I am at a loss how to provide a deprecation path for this.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by jklymak — APPROVED** [review]
I can't imagine anyone will complain if this doesn't get a deprecation.

**Comment by rcomer** [comment]
I belatedly found the [plot_children](https://github.com/matplotlib/matplotlib/blob/6166e5b990b32c7f46847782c0354653e483560d/lib/matplotlib/_layoutgrid.py#L500) function so can now visualise that I'm using the blue rectangle to position the title.

test

File: lib/matplotlib/_constrained_layout.py [hunk 1]
```diff
@@ -140,6 +140,13 @@ def do_constrained_layout(fig, h_pad, w_pad,
                                     w_pad=w_pad, hspace=hspace, wspace=wspace)
                 else:
                     _api.warn_external(warn_collapsed)
+
+                if ((suptitle := fig._suptitle) is not None and
+                        suptitle.get_in_layout() and suptitle._autopos):
+                    x, _ = suptitle.get_position()
+                    suptitle.set_position(
+                        (x, layoutgrids[fig].get_inner_bbox().y1 + 0.02))
```
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:148]: Where does 0.02 come from? Does it work well with different font sizes?

Comment (rcomer) [reply lib/matplotlib/_constrained_layout.py:148]: 0.02 is the default space between the top of this title and the top of the figure without constrained layout, so I chose it to be consistent with that.

Comment (rcomer) [reply lib/matplotlib/_constrained_layout.py:148]: I think all these look OK

test_xx-large
test_x-large
test_large
test_medium
test_small

Comment (rcomer) [reply lib/matplotlib/_constrained_layout.py:148]: Here's what happens with my current implementation if compressed layout is not moving axes in the vertical and I do not use `bbox_inches='tight'`.  I will change it to use _h_pad_.

test_x-large(1)

Comment (rcomer) [reply lib/matplotlib/_constrained_layout.py]: With the update, here is what I get with `figsize=(5, 5)` and no `bbox_inches='tight'`:
test_x-large


Here is what I get with `figsize=(4, 10)` and saving with `bbox_inches='tight', pad_inches='layout'`:
test_x-large(1)

**Comment by rcomer** [comment]
Docs are broken on `main`.

**Review by jklymak — APPROVED** [review]
Still looks good to me!

**Review by timhoffm — APPROVED** [review]

**Pull request merged**